### DISCUSSION TITLE
Use Robolectric to test whether Java library works on Android API version x?

### DISCUSSION THREAD

**Discussion by p-bakker** [robolectric/robolectric#7771] [Q&A] [answered]
Hi,

I'm involved in a Java project that some people use ice Android.

The project has an extensive unit test suite and we're looking for a way to run that existing test suite against one or more Android (api) versions, to see whether our Java code is compatible with that specific Android API version (to surface issues due to changes in our code that are incompatible with a certain Android API version)

Not being will versed in Android Android myself, I'm wondering whether Robolectric would fit this usecase.

Additional requirement: need to be able to run the testsuite on a CI server (we use GitHub Actions)

Hoping for some insight,
P.

**Comment by utzcoz** [comment]
Hi @p-bakker ,

> Not being will versed in Android Android myself, I'm wondering whether Robolectric would fit this usecase.

IMO, Robolectric could be your choice to try. Robolectric is a suite of Java library that simulate Android environment on pure JVM. You can use it to test your library with different Android API/SDK versions, like Robolectric it self does. Robolectric's shadows simulate a large suite of Android APIs with great fidelity, for example AMS, WMS, PMS, and you can trust it like other large open-source projects like Flutter, Firebase-Android-SDK, AOSP, Chromium. Also we can improve some APIs' fidelity of Robolectric if you think they can be better.

> Additional requirement: need to be able to run the testsuite on a CI server (we use GitHub Actions)

Robolectric can meet this requirement because Robolectric runs its large test suite on GitHub Actions with pure Gradle command to test every commit. You can create some sample Android [library] projects that integrate your library and Roblectric, and then run `./gradlew test` to run all your tests on your local machine and GitHub Actions/other CI environment.

**Comment by p-bakker** [comment]
@utzcoz thnx for your rapid response!

> Robolectric's shadows simulate a large suite of Android APIs with great fidelity, for example AMS, WMS, PMS

That part is not of particular interest to us, as our library doesn't use any of that, as our library is a generic Java library

What we are interested in is having caught that for example `javax.lang.model.SourceVersion` is not available on Android or that `java.util.Map.putIfAbsent` is not available on Android api version 21.

Given that Robolectric runs inside the JVM, would such issues surface by throwing NoClassDefFoundError/NoSuchMethodError?

**Reply by utzcoz** [reply]
> What we are interested in is having caught that for example javax.lang.model.SourceVersion is not available on Android or that java.util.Map.putIfAbsent is not available on Android api version 21.

From my knowledge, it might be not. Because the suite runs in JVM, and its classpath contians JVM libraries that might contain these APIs.   It might need @hoisie 's confirmation.

**Comment by hoisie** [comment, marked as answer]
> What we are interested in is having caught that for example javax.lang.model.SourceVersion is not available on Android or that java.util.Map.putIfAbsent is not available on Android api version 21.

Robolectric can detect incompatibilities in Android API but not Java APIs. I think these days you need at least Java 9 TO run Robolectric. If you want more comprehensive compatibility tests that can check Java incompatibilities I'd recommend emulator tests. In theory there may be something you can do with custom instrumentation but it may be complex.

**Reply by utzcoz** [reply]
@p-bakker I also recommend Emulator tests for your purpose as hoisie said. If you use GitHub Actions, you can check what Robolectric uses for Emulator tests on CI.  If you use custom CI, you can check https://github.com/google/android-emulator-container-scripts, and it can help you to run headless Emulator on your container.

**Comment by p-bakker** [comment]
Thnx for all the (quick) responses and answers.

To bad Robolectric isn't a match for us, looked like it otherwise could've been a pretty straight-forward solution to our challenge.

I'll go on and look at the emulator route, although from what I've read, these have their challenges as well, especially running them in a CI environment in a performant way.

Wish one could just launch a virtual VM that just has the Dalvik (or whatever it's successors are called) environment, as for our purpose we don't need a full emulated environment
